### PR TITLE
Added check to `forc_abi` to fail on `non-contracts`

### LINKED ISSUE #526: Forc ABI generation doesn't fail for scripts
Forc version: v0.1.6

Running `forc json-abi` for a `script` project gives the following output instead of an error:

```console
$ forc json-abi
  Compiled library "core".
  Compiled library "core".
  Compiled library "lib-std".
  Compiled script "foo".
[]
```

Only contracts have an ABI, so `forc json-abi` should fail with an explanative error otherwise.

### LINKED ISSUE #1106: Remove `forc::utils::CliError` in favour of `anyhow::Result`
Right now `forc`'s `CliError` is basically a limited re-implementation of a subset of `anyhow`'s error type. We should remove it in favour of using `anyhow` which is already a dependency.

We can keep the error constructor functions (e.g. `manifest_file_missing`, etc) but as standalone functions in the `cli` mod that return `anyhow::Error` instead.

### LINKED ISSUE #1117: `forc`'s test harness checks `non-contract`s for `json abi`
`number_of_tests_run` in `e2e_vm_tests::mod` currently tests all `should_pass` test programs for `json abi`, including `non-contract`s.

### PR DESCRIPTION
**Pull request by eureka-cpu**
Closes #526 -- Initially reused the logic found in `forc_deploy` & `forc_run` to throw an error instead of an empty abi `[]` if the wrong `program` type was found.

Closes #1106 -- Replaced the `CliError` struct with `anyhow::Error`.

Closes #1117 -- Separated the tests in vec `positive_project_names` in `e2e_vm_tests::mod` into `...no_abi`, `...with_abi`, removed the `json.oracle` files and updated the surrounding functions to handle the aforementioned changes.

I've also converted the reused logic in `forc_abi_json`, `forc_deploy` & `forc_run` into a single function called `check_program_type`. To avoid circular dependency, I've moved it and its associated functions out of `utils::cli_error` and into `forc_pkg::pkg` since we decided to implement `program_type` as a method on `manifest::Manifest`. This helps by removing duplicate code which is being tracked here: #1107 .

### CODE REVIEW

File: forc/src/ops/forc_abi_json.rs [hunk 1]
```diff
@@ -1,30 +1,80 @@
...
+                        project_name,
+                        SWAY_CONTRACT,
+                        SWAY_LIBRARY,
+                    )),
+                },
+                None => Err(CliError::parsing_failed(project_name, parsed_result.errors)),
+            }
+        }
+        None => Err(CliError::manifest_file_missing(curr_dir)),
     }
```
Comment (mitchmindtree) [inline forc/src/ops/forc_abi_json.rs:79]: Perhaps it's worth abstracting this into a `pkg::check_is_contract` function (in `forc_pkg`) or something along these lines?

**Review by mitchmindtree — CHANGES_REQUESTED** [review]

File: forc/src/ops/forc_abi_json.rs (hunk 1 above)
Comment (eureka-cpu) [reply forc/src/ops/forc_abi_json.rs:79]: exactly my thoughts, I opened #1107 a few mins ago and added this

File: forc/src/utils/cli_error.rs [hunk 2]
```diff
@@ -48,7 +48,7 @@ impl CliError {
 
 impl fmt::Display for CliError {
     fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
-        write!(f, "{}", self)
+        write!(f, "{:?}", self)
```
[the remark this answers is not in the document]
Comment (eureka-cpu) [reply forc/src/utils/cli_error.rs:51]: I'll take care of this tomorrow morning, thanks!

**Comment by eureka-cpu** [comment]
@adlerjohn & @mitchmindtree getting this error on my machine even though CI passed: `Deploying should_pass/test_contracts/basic_storage
  Compiled library "core".
  Compiled library "std".
  Compiled library "basic_storage_abi".
  Compiled contract "basic_storage".
  Bytecode size is 108 bytes.
Contract id: 0xfa4d1b8c5dcbdd3a4ea073d162cb2313e89a03da03870a5561c12ffe10bca189
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: An error occured while creating a new object: Connection refused (os error 61)', test/src/e2e_vm_tests/harness.rs:31:10`

**Comment by mohammadfawaz** [comment]
> @adlerjohn & @mitchmindtree getting this error on my machine even though CI passed: 
> 
> 
> 
> ```
> 
> Deploying should_pass/test_contracts/basic_storage
> 
>   Compiled library "core".
> 
>   Compiled library "std".
> 
>   Compiled library "basic_storage_abi".
> 
>   Compiled contract "basic_storage".
> 
>   Bytecode size is 108 bytes.
> 
> Contract id: 0xfa4d1b8c5dcbdd3a4ea073d162cb2313e89a03da03870a5561c12ffe10bca189
> 
> thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: An error occured while creating a new object: Connection refused (os error 61)', test/src/e2e_vm_tests/harness.rs:31:10
> 
> ```

Do you have `fuel-core` running? If a, try to clean up ~/.forc and ~./fuel and then relaunch it.

**Comment by eureka-cpu** [comment]
> > @adlerjohn & @mitchmindtree getting this error on my machine even though CI passed:
> > ```
> > 
> > Deploying should_pass/test_contracts/basic_storage
> > 
> >   Compiled library "core".
> > 
> >   Compiled library "std".
> > 
> >   Compiled library "basic_storage_abi".
> > 
> >   Compiled contract "basic_storage".
> > 
> >   Bytecode size is 108 bytes.
> > 
> > Contract id: 0xfa4d1b8c5dcbdd3a4ea073d162cb2313e89a03da03870a5561c12ffe10bca189
> > 
> > thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: An error occured while creating a new object: Connection refused (os error 61)', test/src/e2e_vm_tests/harness.rs:31:10
> > ```
> 
> Do you have `fuel-core` running? If so, try to clean up ~/.forc and ~./fuel and then relaunch it.

Will do, thanks! If all looks well, then disregard my previous comment @adlerjohn & @mitchmindtree

File: forc/src/ops/forc_abi_json.rs (hunk 1 above)
Comment (eureka-cpu) [reply forc/src/ops/forc_abi_json.rs:40]: For this PR, I just kept it all inside `cli_error.rs`, let me know if this needs to change

**Comment by adlerjohn** [comment]
Is this ready for review?

**Comment by eureka-cpu** [comment]
> Is this ready for review?

yes!

**Review by adlerjohn — COMMENTED** [review]
Should you remove the JSON ABI oracle files for scripts in the test suite?

**Comment by eureka-cpu** [comment]
>

**Comment by eureka-cpu** [comment]
> Should you remove the JSON ABI oracle files for scripts in the test suite?

accidentally closed while trying to quote reply 🤦🏻 yes, I'll remove those rn thank you!

File: forc/src/utils/cli_error.rs [hunk 3]
```diff
@@ -1,91 +1,67 @@
...
+    Error::msg(message)
 }
 
-impl From<&str> for CliError {
-    fn from(s: &str) -> Self {
-        CliError {
-            message: s.to_string(),
-        }
-    }
+pub fn wrong_sway_type(project_name: &str, wanted_type: &str, parse_type: &str) -> anyhow::Error {
```
Comment (sezna) [inline forc/src/utils/cli_error.rs:28]:
```suggestion
pub fn wrong_project_type(project_name: &str, wanted_type: &str, parse_type: &str) -> anyhow::Error {
```
Or
```suggestion
pub fn wrong_project_type(project_name: &str, wanted_type: &str, parse_type: &str) -> anyhow::Error {
```
I think those names would both be more descriptive.

File: sway-core/src/semantic_analysis/syntax_tree.rs [hunk 4]
```diff
@@ -15,7 +15,9 @@ use crate::{
     type_engine::*,
     AstNode, ParseTree,
 };
-
+use std::fmt::Display;
+use std::fmt::Formatter;
+use std::fmt::Result;
```
Comment (sezna) [inline sway-core/src/semantic_analysis/syntax_tree.rs:20]:
```suggestion
use std::fmt::{Display, Formatter, Result};
```

File: test/src/e2e_vm_tests/mod.rs [hunk 5]
```diff
@@ -288,20 +279,51 @@ pub fn run(filter_regex: Option<regex::Regex>) {
...
+        // contracts revert because this test runs them against the VM
+        // and no selectors will match
+        (
+            "should_pass/test_contracts/contract_abi_impl",
+            ProgramState::Revert(0),
+        ),
+        (
+            "should_pass/language/valid_impurity",
+            ProgramState::Revert(0), // false
+        ),
```
Comment (sezna) [inline test/src/e2e_vm_tests/mod.rs:308]: All test contracts should have their ABI checked -- just ran a quick grep for them.
```console
test_contracts/basic_storage/src/main.sw
1:contract;

test_contracts/test_fuel_coin_contract/src/main.sw
1:contract;

test_contracts/increment_contract/src/main.sw
1:contract;

test_contracts/auth_testing_contract/src/main.sw
1:contract;

test_contracts/balance_test_contract/src/main.sw
1:contract;

test_contracts/storage_access_contract/src/main.sw
1:contract;

test_contracts/context_testing_contract/src/main.sw
1:contract;

test_contracts/contract_abi_impl/src/main.sw
1:contract;

language/valid_impurity/src/main.sw
1:contract;

language/storage_declaration/src/main.sw
1:contract;
```

File: test/src/e2e_vm_tests/mod.rs [hunk 6]
```diff
@@ -288,20 +279,51 @@ pub fn run(filter_regex: Option<regex::Regex>) {
...
+        .iter()
+        .fold(0, |acc, (name, res)| {
+            if filter(name) {
+                assert_eq!(crate::e2e_vm_tests::harness::runs_in_vm(name), *res);
+                // cannot use partial eq on type `anyhow::Error` so I've used `matches!` here instead.
+                // https://users.rust-lang.org/t/issues-in-asserting-result/61198/3 for reference.
+                assert!(matches!(
+                    crate::e2e_vm_tests::harness::test_json_abi(name),
+                    Ok(())
+                ));
```
Comment (sezna) [inline test/src/e2e_vm_tests/mod.rs:321]:
```suggestion
                assert!
                    crate::e2e_vm_tests::harness::test_json_abi(name).is_ok()
                );
```

We can't use this because it is nightly, but just in case you were curious, there's [assert_match](https://doc.rust-lang.org/std/assert_matches/macro.assert_matches.html) coming soon (TM).

**Review by sezna — CHANGES_REQUESTED** [review]

File: forc/src/utils/cli_error.rs (hunk 3 above)
Comment (otrho) [reply forc/src/utils/cli_error.rs:28]: You've suggested `wrong_project_type` twice? :slightly_smiling_face:

File: forc/src/cli/commands/run.rs [hunk 7]
```diff
@@ -89,7 +89,7 @@ pub struct Command {
 
 pub(crate) async fn exec(command: Command) -> Result<()> {
     match forc_run::run(command).await {
-        Err(e) => bail!(e.message),
```
[the remark this answers is not in the document]
Comment (eureka-cpu) [reply forc/src/cli/commands/run.rs:92]: `message` was from the `CliError` struct. I changed it since `message` no longer exists, but we want to retain the error.

File: forc/src/ops/forc_deploy.rs [hunk 8]
```diff
@@ -32,73 +30,40 @@ pub async fn deploy(command: DeployCommand) -> Result<fuel_tx::ContractId, CliEr
...
-                },
-                None => Err(CliError::parsing_failed(project_name, parsed_result.errors)),
-            }
+    match client.submit(&tx).await {
+        Ok(logs) => {
+            println!("Logs:\n{:?}", logs);
+            Ok(contract_id)
         }
-        None => Err(CliError::manifest_file_missing(curr_dir)),
+        Err(e) => Err(anyhow!("{}", e)),
```
[the remark this answers is not in the document]
Comment (eureka-cpu) [reply forc/src/ops/forc_deploy.rs:66]: Not sure why it's lined up this way, otherwise you would think that. This is actually from L79 `Err(e) => Err(e.to_string().into())`

File: forc/src/ops/forc_deploy.rs [hunk 9]
```diff
@@ -32,73 +30,40 @@ pub async fn deploy(command: DeployCommand) -> Result<fuel_tx::ContractId, CliEr
...
+    );
 
-                        let node_url = match &manifest.network {
-                            Some(network) => &network.url,
-                            _ => DEFAULT_NODE_URL,
-                        };
+    let node_url = match &manifest.network {
+        Some(network) => &network.url,
+        _ => DEFAULT_NODE_URL,
+    };
```
[the remark this answers is not in the document]
Comment (eureka-cpu) [reply forc/src/ops/forc_deploy.rs:57]: The indentation came from the code bloat, I just organized it after I reduced it down to the stored variable `manifest` on L17. `let manifest = check_tree_type(curr_dir, SWAY_CONTRACT)?;`

File: forc/src/utils/cli_error.rs (hunk 3 above)
Comment (eureka-cpu) [reply forc/src/utils/cli_error.rs:28]: I agree with both versions 🙂

File: test/src/e2e_vm_tests/mod.rs (hunk 5 above)
Comment (eureka-cpu) [reply test/src/e2e_vm_tests/mod.rs:308]: In this PR I just took what was already there in the `positive_project_names` and separated them to make the test that was previously there, pass.

File: test/src/e2e_vm_tests/mod.rs (hunk 6 above)
Comment (eureka-cpu) [reply test/src/e2e_vm_tests/mod.rs:321]: I did see that as an option, really looking forward to some of the nightly options!

File: test/src/e2e_vm_tests/mod.rs (hunk 5 above)
Comment (eureka-cpu) [reply test/src/e2e_vm_tests/mod.rs:308]: Opened #1124 since this PR already encompasses so much

File: forc/src/utils/cli_error.rs (hunk 3 above)
Comment (sezna) [reply forc/src/utils/cli_error.rs]: Oh I meant wrong tree type or wrong project type. My bad!

File: test/src/e2e_vm_tests/mod.rs (hunk 6 above)
Comment (sezna) [reply test/src/e2e_vm_tests/mod.rs:321]: BTW, my suggestion is still valid on stable. It doesn't use `assert_matches`:
```rust
        assert!
                    crate::e2e_vm_tests::harness::test_json_abi(name).is_ok()
                );
```

File: test/src/e2e_vm_tests/mod.rs (hunk 5 above)
Comment (sezna) [reply test/src/e2e_vm_tests/mod.rs:308]: Contracts don't actually run and return a value, so they all have a constant return state. It is invariable. 

We need to test each of their ABIs though -- before this PR, we were checking all of those ABIs via the ABI tests. This PR removed  the ABI tests for _everything_ and only re-added them for two contracts, when in reality there are quite a few more. I think we should be able to keep those ABI tests in this PR, otherwise we would be losing test coverage.

Comment (eureka-cpu) [reply test/src/e2e_vm_tests/mod.rs:308]: >This PR removed the ABI tests for everything and only re-added them for two contracts, when in reality there are quite a few more. 

I only separated the ABI tests that were a part of `positive_project_names`, the rest of the contracts listed here didn't change. If they aren't being tested now, they weren't being tested before?

**Review by mitchmindtree — CHANGES_REQUESTED** [review]

File: forc/src/utils/cli_error.rs [hunk 10]
```diff
@@ -1,91 +1,71 @@
...
-
-impl From<AddrParseError> for CliError {
-    fn from(e: AddrParseError) -> Self {
-        CliError {
-            message: e.to_string(),
-        }
-    }
-}
+/// Given the current directory and file type, determines whether the correct project type is present.
+pub fn check_project_type(curr_dir: PathBuf, wanted_type: &str) -> Result<Manifest> {
```
[the remark this answers is not in the document]
Comment (eureka-cpu) [reply forc/src/utils/cli_error.rs:46]: I actually found a way to make this work with `TreeType`, it does end up giving you _more_ info if you happen to give it a Library, but I actually think that's cleaner than recreating a second enum with essentially the same properties (feel free to dispute that), another problem with that is that we would be trying to match two different types. One of `TreeType` and one of `ProgramType`.

[the remark this answers is not in the document]
Comment (eureka-cpu) [reply forc/src/utils/cli_error.rs]: To avoid recreating the `manifest` so many times I've gone ahead and created something of a compromise from your above code and my previous code. I think this feels much nicer than before, I also moved the related functions from `cli_error` to `forc-pkg::pkg` as you previously instructed.

File: forc-pkg/src/pkg.rs [hunk 11]
```diff
@@ -1171,3 +1171,62 @@ fn test_source_git_pinned_parsing() {
...
+}
+
+/// Format an error message if a given URL fails to produce a working node.
+pub fn fuel_core_not_running(node_url: &str) -> anyhow::Error {
+    let message = format!("could not get a response from node at the URL {}. Start a node with `fuel-core`. See https://github.com/FuelLabs/fuel-core#running for more information", node_url);
+    Error::msg(message)
+}
+
+/// Given the current directory and expected program type, determines whether the correct program type is present and returns a manifest.
+pub fn check_program_type(curr_dir: PathBuf, expected_type: TreeType) -> Result<Manifest> {
```
Comment (mitchmindtree) [inline forc-pkg/src/pkg.rs:1216]: I realise that for a couple of the upstream cases (`forc_deploy`, `forc_run`) that creating the `Manifest` inside this function saves a couple extra lines in those places. However, I think there's a bit of a conflation of responsibilities in having this function return a `Manifest`, and I'm not sure it's worth saving a couple lines downstream.

I still think passing in a reference to an existing `Manifest` is preferable as it keeps the separation of responsibilities clear. It also makes this function a little more modular for future use, e.g. if we come across a case where we've already loaded a `Manifest` but want to check the program type, we should not need to have to do more I/O and deserialize again just to check the program type. I hope this makes sense!

File: forc/src/ops/forc_abi_json.rs [hunk 12]
```diff
@@ -1,18 +1,30 @@
...
         path: command.path,
         offline_mode: command.offline_mode,
         silent_mode: command.silent_mode,
         minify_json_abi: command.minify,
         ..Default::default()
     };
+
     let compiled = crate::ops::forc_build::build(build_command)?;
     let json_abi = json!(compiled.json_abi);
+
```
Comment (mitchmindtree) [inline forc/src/ops/forc_abi_json.rs:27]: Not important, but try to avoid adding newlines, touching whitespace or formatting unrelated code as it can cause a bit of git diff noise and occasionally cause unnecessary conflicts.

Don't worry, I'm guilty of accidentally removing trailing whitespace in random places because of my editor setup sometimes :face_with_spiral_eyes:

File: forc-pkg/src/pkg.rs (hunk 11 above)
Comment (mitchmindtree) [reply forc-pkg/src/pkg.rs:1216]: Perhaps a clearer alternative might be:

```rust
impl Manifest {
    // ...

    //// The same as `Manifest::from_dir`, but checks the project is of the given program type before returning.
    pub fn from_dir_expecting_program_type(dir: &Path, ty: &TreeType) -> Result<Self> {
        // Implemented in terms of `Manifest::from_dir` and `check_program_type`.
    }
```

I think as a part of this, we could update the `Manifest::from_dir` function to automatically check parent directories in the case that the immediate directory doesn't contain a manifest. This might save us from always having to call `find_manifest_dir` everywhere before we can call `Manifest::from_dir`.

**Review by mitchmindtree — CHANGES_REQUESTED** [review]
Getting closer!

File: test/src/e2e_vm_tests/mod.rs (hunk 5 above)
Comment (sezna) [reply test/src/e2e_vm_tests/mod.rs:308]: Ah, I see. This is actually a significant test gap and regression in test coverage that previously went undetected. We have not been testing the ABIs of anything that actually _has_ an ABI...eh....

Ok, could you tackle #1124 after this in a fast-follow PR then? That's a pretty important test to have.

File: forc-pkg/src/pkg.rs (hunk 11 above)
Comment (eureka-cpu) [reply forc-pkg/src/pkg.rs:1216]: >I think as a part of this, we could update the Manifest::from_dir function to automatically check parent directories in the case that the immediate directory doesn't contain a manifest. This might save us from always having to call find_manifest_dir everywhere before we can call Manifest::from_dir.
I do like that idea, would you mind if I went ahead and made these edits and then tackled this in another PR?

Comment (mitchmindtree) [reply forc-pkg/src/pkg.rs:1216]: Yeah that sounds fine to me!

File: forc-pkg/src/manifest.rs [hunk 13]
```diff
@@ -150,6 +153,18 @@ impl Manifest {
             Dependency::Simple(_) => None,
         })
     }
+
+    /// Parse and return the associated project's program type.
+    pub fn program_type(&self, manifest_dir: PathBuf) -> Result<TreeType> {
+        let manifest = Manifest::from_dir(&manifest_dir)?;
```
Comment (mitchmindtree) [inline forc-pkg/src/manifest.rs:159]: Ahh I missed this previously - there's no need to load a new `Manifest` in this method, we should use `self` instead.

**Review by mitchmindtree — CHANGES_REQUESTED** [review]

File: forc-pkg/src/manifest.rs (hunk 13 above)
Comment (eureka-cpu) [reply forc-pkg/src/manifest.rs:159]: Like so? 

```pub fn program_type(&self, manifest_dir: PathBuf) -> Result<TreeType> {
        let entry_string = self.entry_string(&manifest_dir)?;
        let program_type = parse(entry_string, None);

        match program_type.value {
            Some(parse_tree) => Ok(parse_tree.tree_type),
            None => bail!(parsing_failed(&self.project.name, program_type.errors)),
        }
    }
```

**Review by sezna — APPROVED** [review]
approving to dismiss my requested changes but deferring to @mitchmindtree

**Review by mitchmindtree — APPROVED** [review]
Epic work @eureka-cpu, I think this is good to go!
